go/vt/vtadmin/cluster/cluster.go

```diff
@@ -373,50 +374,269 @@ func (c *Cluster) getTablets(ctx context.Context) ([]*vtadminpb.Tablet, error) {
+// filtering of either this set, or the provided Tablets, will happen later.
+//
+// (2) If opts.SizeOpts.AggregateSizes is true, we will also make a call to
+// FindAllShardsInKeyspace, in order to fan out GetSchema RPCs to a tablet in
+// each shard. If this option is false, we make exactly one GetSchema request to
+// a single, randomly-chosen, tablet in the keyspace.
 //
-// Note that the request's TabletAlias field will be ignored, using the provided
-// tablet's Alias instead. This override is done on a copy of the request, so it
-// is transparent to the caller.
+// (2.1) If, in size aggregation mode, opts.SizeOpts.IncludeNonServingShards is
+// false (the default), then we will filter out any shards for which
+// IsMasterServing is false in the topo, and make GetSchema RPCs to one tablet
+// in every _serving_ shard. Otherwise we will make a GetSchema RPC to one
+// tablet in _every_ shard.
 //
-// This function takes both the request argument and tablet argument to
-// (a) set the Keyspace field on the resulting Schema object, which comes from
-// the provided tablet; and, (b) allow a caller, like vtadmin.API to collect a
-// bunch of tablets once and make a series of GetSchema calls without Cluster
-// refetching the tablet list each time.
-func (c *Cluster) GetSchema(ctx context.Context, req *vtctldatapb.GetSchemaRequest, tablet *vtadminpb.Tablet) (*vtadminpb.Schema, error) {
+// (3) Irrespective of whether we're including nonserving shards, or whether
+// we're doing size aggregation at all, we will only make GetSchema RPCs to
+// tablets that are in SERVING state; we don't want to use a tablet that might
+// be in a bad state as the source of truth for a schema. Therefore if we can't
+// find a SERVING tablet for the keyspace (in non-aggregation mode) or for a
+// shard in that keyspace (in aggregation mode), then we will return an error
+// back to the caller.
+func (c *Cluster) GetSchema(ctx context.Context, keyspace string, opts GetSchemaOptions) (*vtadminpb.Schema, error) {
 	span, ctx := trace.NewSpan(ctx, "Cluster.GetSchema")
 	defer span.Finish()
 
+	if opts.SizeOpts == nil {
+		opts.SizeOpts = &vtadminpb.GetSchemaTableSizeOptions{
+			AggregateSizes:          false,
+			IncludeNonServingShards: false,
+		}
+	}
+
+	if opts.BaseRequest == nil {
+		opts.BaseRequest = &vtctldatapb.GetSchemaRequest{}
+	}
+
+	if opts.SizeOpts.AggregateSizes && opts.BaseRequest.TableNamesOnly {
+		log.Warningf("GetSchema(cluster = %s) size aggregation is incompatible with TableNamesOnly, ignoring the latter in favor of aggregating sizes", c.ID)
+		opts.BaseRequest.TableNamesOnly = false
+	}
+
 	AnnotateSpan(c, span)
+	span.Annotate("keyspace", keyspace)
+	annotateGetSchemaRequest(opts.BaseRequest, span)
+	vtadminproto.AnnotateSpanWithGetSchemaTableSizeOptions(opts.SizeOpts, span)
 
-	// Copy the request to not mutate the caller's request object.
-	r := *req
-	r.TabletAlias = tablet.Tablet.Alias
+	if len(opts.Tablets) == 0 {
```

Comment:
I think this whole block could be simplify if we refactor into smaller functions. An idea that comes to mind is:
* One function for the case tablets == 0 (let's call it X for now)
* One function for the else case (let's call it Y)
* One function that does all the actual work and queries the tablets (let's call Z). Then I could this block be more like:
```
if len(opts.Tablets) == 0 {
  tabletsToQuery := X(opts)
  return Z(tabletsToQuery)
}

tabletsToQuery := Y(opts)
return Z(tabletsToQuery)
```

Something like that.

Reply:
I like this a lot, it should be much cleaner to read (I agree this function is suuuuper bulky haha)